R/tokens_xptr.R

```diff
@@ -0,0 +1,209 @@
+#' @rdname tokens_xptr
+#' @export
+is.tokens_xptr <- function(x) {
+    identical(typeof(x), "externalptr") && "tokens_xptr" %in% class(x)
+}
+
+#' Convert a tokens object to a tokens_xptr object
+#'
+#' tokens_xptr is a external pointer object to process a large number of
+#' documents efficiently.
+#' @param x a [tokens] object to convert or a tokens_xptr object to
+#'   to deep-copy.
+#' @rdname tokens_xptr
+#' @export
+as.tokens_xptr <- function(x) {
+    UseMethod("as.tokens_xptr")
+}
+
+#' @rdname tokens_xptr
+#' @method as.tokens_xptr tokens
+#' @export
+as.tokens_xptr.tokens <- function(x) {
+    attrs <- attributes(x)
+    result <- cpp_as_xptr(x, attrs$types)
+    build_tokens(result, 
+                 types = NULL, 
+                 padding = TRUE, 
+                 docvars = attrs[["docvars"]], 
+                 meta = attrs[["meta"]], 
+                 class = attrs[["class"]])
+}
+
+#' @rdname tokens_xptr
+#' @method as.tokens_xptr tokens_xptr
+#' @export
+as.tokens_xptr.tokens_xptr <- function(x) {
+    attrs <- attributes(x)
+    result <- cpp_copy_xptr(x)
+    rebuild_tokens(result, attrs)
+}
+
+#' @export
+ndoc.tokens_xptr <- function(x) {
+    cpp_ndoc(x)
+}
+
+#' @export
+types.tokens_xptr <- function(x) {
+    cpp_get_types(x, TRUE)
+}
+
+#' @export
+ntoken.tokens_xptr <- function(x, ...) {
+    structure(cpp_ntoken(x), names = docnames(x))
+}
+
+# #' @export
+# docnames.tokens_xptr <- function(x) {
+#     get_docvars(x, "docname_", FALSE, TRUE, TRUE)
+# }
+# 
+# #' @export
+# docid.tokens_xptr <- function(x) {
+#     get_docvars(x, "docid_", FALSE, TRUE, TRUE)
+# }
+# 
+# #' @export
+# segid.tokens_xptr <- function(x) {
+#     get_docvars(x, "segid_", FALSE, TRUE, TRUE)
+# }
+
+#' @method as.tokens tokens_xptr
+#' @export
+as.tokens.tokens_xptr <- function(x, ...) {
```

Comment:
Should this (and other functions) have a `concatenator` argument, as do the other `as.tokens()` methods?

Reply:
Unnecessary because `x` can only be `tokens`. Conversion of a list will be `as.tokens_xptr(as.tokens(list(), concatenator = "+"))`.